### DISCUSSION TITLE
Wayland / Gtk3 support?

### DISCUSSION THREAD

**Discussion by voxik** [doublecmd/doublecmd#1097] [Suggestions]
### Description

With Gnome 44, the Double Commander running via XWayland, the windows are not correctly placed / remembered and what not. I assume this is mutter issue. But one solution to this would be if DC supported Wayland. So I though I'll give it a shot and try if Gtk3 is the way to go:

~~~
$ lcl=gtk3 ./build.sh beta

... snip ...

(3104) Compiling ./frames/ucomponentssignature.pas
(3104) Compiling uglobs.pas
(3104) Compiling ucolumns.pas
(3104) Compiling uglobs.pas
(3104) Compiling uhotkeymanager.pas
(3104) Compiling ./platform/ukeyboard.pas
/builddir/build/BUILD/doublecmd-1.0.11/src/./platform/ukeyboard.pas(228,1) Fatal: (2003) Syntax error, "identifier" expected but "VAR" found
Fatal: (1018) Compilation aborted
Error: /usr/bin/ppcx64 returned an error exitcode
Error: (lazarus) Compile Project, Mode: Release, Target: /builddir/build/BUILD/doublecmd-1.0.11/doublecmd: stopped with exit code 1
Error: (lazbuild) failed compiling of project /builddir/build/BUILD/doublecmd-1.0.11/src/doublecmd.lpi
~~~

So it seems it got quite far with compilation prior it failed.

Not really knowing what I am doing and not spending too much time with investigation, I naively tweaked the file a bit:

~~~diff
$ git diff
diff --git a/src/platform/ukeyboard.pas b/src/platform/ukeyboard.pas
index 8b6a4426f..170834019 100644
--- a/src/platform/ukeyboard.pas
+++ b/src/platform/ukeyboard.pas
@@ -157,6 +157,8 @@ uses
   , Gdk2, GLib2, Gtk2Extra
   , Gtk2Proc
 {$ENDIF}
+  , LazGdk3, LazGLib2
+  , Gtk3Procs
 {$IF DEFINED(X11) and (DEFINED(LCLQT) or DEFINED(LCLQT5))}
   {$IF DEFINED(LCLQT)}
   , qt4, qtwidgets
@@ -190,16 +192,18 @@ var
   XDisplay: PDisplay = nil;
   {$ELSEIF DEFINED(LCLGTK2)}
   XDisplay: PGdkDisplay = nil;
+  {$ELSEIF DEFINED(LCLGTK3)}
+  XDisplay: PGdkDisplay = nil;
   {$ELSEIF (DEFINED(LCLQT) or DEFINED(LCLQT5))}
   XDisplay: PDisplay = nil;
   {$ENDIF}
 {$ENDIF}
 
-{$IF DEFINED(UNIX) and (DEFINED(LCLGTK) or DEFINED(LCLGTK2))}
+{$IF DEFINED(UNIX) and (DEFINED(LCLGTK) or DEFINED(LCLGTK2) or DEFINED(LCLGTK3))}
 var
   // This is set to a virtual key number that AltGr is mapped on.
   VK_ALTGR: Byte = VK_UNDEFINED;
-  {$IF DEFINED(LCLGTK2)}
+  {$IF DEFINED(LCLGTK2) or DEFINED(LCLGTK3)}
   KeysChangesSignalHandlerId : gulong = 0;
   {$ENDIF}
 {$ENDIF}
~~~

This helped with the compilation and with some error, resulted in kind of "running" Double Commander:

![obrazek](https://user-images.githubusercontent.com/14406/230767854-9a8aac74-8af3-46b7-9f5b-12547b69ca72.png)

Well, it does not really work, but if nothing else, it now runs native Wayland.

So I wonder, is the Gtk3 support something what was considered and would be feasible?

**Comment by alexx2000** [comment]
Use Double Commander Qt5 version if you need native Wayland support.

Lazarus Gtk3 support has alpha state so it's unusable. I doubt that Gtk3 support will appear in the coming years.

You [can see](https://gitlab.com/freepascal.org/lazarus/lazarus/-/commits/main/lcl/interfaces/gtk3) that no active work on Gtk3 support.

**Comment by gorhatlin** [comment]
It seems like GTK3 development is very active lately, even Arch Linux has lazarus-gtk3 package in their repo which seems quite stable. Could we still get GTK3 version of our beloved Double Commander?

**Comment by alexx2000** [comment]
Looks like Lazarus 5.0 will have a normal Gtk3 support.

**Comment by alexx2000** [comment]
I added experimental GTK3 builds using Lazarus 4.99 to [snapshots](https://github.com/doublecmd/snapshots/releases/latest), requires Linux >= 2022.04.

**Reply by advice2020** [reply]
What are you doing to me Alexander!!!!
The Linux toolkit nightmares I had been experiencing for years now finally went away and now you do this! A GTK3 version!!!

:)
Seriously, thank you very much for taking the time to work on this.
I really wanted this in the past before I went on my QT journey with DC.
I did not realize that it was Lazarus that was preventing a GTK3 version release.
In the past I had just thought that QT was going to be the only toolkit to be used for DC in the future.

Where did you even get this version of Lazarus? 4.99.0.0!
You mentioned in past that you use Arch but even still.
Debian 13 stable = 4.0.0
Debian 13 SID = 4.6.0.0
I did not even see this on the Lazarus site itself
Lazarus site = 4.6.0.0

Also, thank you for providing standalone version for users to try.
I would really like to try and help test this out but being a Debian based user it is a little bit of a concern that this version will not be available any time soon for its users, just in case for some reason I needed to compile it.  And in typical Debian fashion they always seem to provide a couple versions just before the one I want when they make their official releases.
Debian 14 will probably get 4.98.0.0 :)

I have already tested it out a little and I found a couple possible issues.
I will post them in the Issues section, hopefully that is okay, let me know if Discussions section is preferred (due to it being in alpha stage).
Depending how things go, I think I am going to try to use it for actual use, this way I could really test it more.

Thank you again for working on this GTK3 versions and for all the work you put into DC in general, really appreciated, this application continues to amaze me.

**Comment by alexx2000** [comment]
DC itself mostly is toolkit independent. Only small part of code is toolkit specific. So most Double Commander GTK3 bugs really are Lazarus bugs.

> Where did you even get this version of Lazarus? 4.99.0.0!

Built from source code: https://gitlab.com/freepascal.org/lazarus/lazarus

**Reply by advice2020** [reply]
Thank you for your response.
Wow, obviously I do not know much about their release schedule but  I wouldn't have thought that their would have been that much of a "gap" between their currently released version 4.6 and their current source code 4.99
Anyways, thank you again for taking the time to compile this bleeding edge source code and producing a version of DC GTK3 with it for some of us to play around with.

**Reply by alexx2000** [reply]
Version 4.99 is the upcoming major 5.0 release. While 4.6 is only a bugfix of 4.0.
